### PR TITLE
v4.0.x: UCX: initialize cuda from ucx pml component

### PR DESCRIPTION
**Pull request by bureddy**
Signed-off-by: Devendar Bureddy <devendar@mellanox.com>

### CODE REVIEW

**Comment by awlauria** [comment]
Please add the -x to the cherry-pick, so we know what master commit it came from. Thanks!

**Comment by gpaulsen** [comment]
@yosefe Please review.

We're curious about what use-case this solves.

**Comment by bureddy** [comment]
@gpaulsen yossie reviewed these changes in master branch (https://github.com/open-mpi/ompi/pull/7893)

when using cuda with UCX, it depends on the cuda runtime enable in openmpi for cuda pack/unpack and datatype copy in collectives codepath.  currently. OMPI cuda runtime is enabled ( mca_common_cuda_stage_one_init()) only if one of the cuda aware BTL(tcp, smcuda or openib) component is opened. 
For example, will see following kind of crash with (`-mca pml ucx -mca btl ^tcp,smcuda,openib`) for `osu_alltoall -d cuda`
```

 0x000000000004f691 non_overlap_copy_content_same_ddt()  /build-result/src/hpcx-v2.6.1-gcc-MLNX_OFED_LINUX-5.0-2.1.8.1.13-redhat7.6-x86_64/ompi-8b4a8cd/opal/datatype/opal_datatype_copy.h:152
 3 0x000000000005dab3 ompi_datatype_sndrcv()  /build-result/src/hpcx-v2.6.1-gcc-MLNX_OFED_LINUX-5.0-2.1.8.1.13-redhat7.6-x86_64/ompi-8b4a8cd/ompi/datatype/ompi_datatype_sndrcv.c:62
 4 0x000000000005ebb1 PMPI_Alltoall()  /build-result/src/hpcx-v2.6.1-gcc-MLNX_OFED_LINUX-5.0-2.1.8.1.13-redhat7.6-x86_64/ompi-8b4a8cd/ompi/mpi/c/profile/palltoall.c:110
 5 0x0000000000401b86 main()  /build-result/src/hpcx-v2.6.1-gcc-MLNX_OFED_LINUX-5.0-2.1.8.1.13-redhat7.6-x86_64/ompi-8b4a8cd/cuda-tests/osu-micro-benchmarks-5.6.2/mpi/collective/osu_alltoall.c:100
 6 0x00000000000223d5 __libc_start_main()  ???:0
```

This PR is considering UCX is cuda aware and enabling OMPI cuda runtime when underlaying UCX is cuda aware.

**Comment by yosefe** [comment]
@gpaulsen OMPI DDT engine expects that PML/BTL which use cuda will call cuda stage1 initialization - otherwise the DDT engine would not try to detect memory type.

**Comment by gpaulsen** [comment]
I don't see a way to just request retest of ARMv8, so here we go:
bot:ompi:retest

**Comment by hppritcha** [comment]
@yosefe please if you want this in to 4.0.5 - we're planning to release in the next several weeks.

File: ompi/mca/pml/ucx/pml_ucx.c [hunk 1]
```diff
@@ -227,22 +230,37 @@ int mca_pml_ucx_open(void)
...
         ucp_cleanup(ompi_pml_ucx.ucp_context);
         ompi_pml_ucx.ucp_context = NULL;
         return OMPI_ERROR;
     }
 
-    ompi_pml_ucx.request_size = attr.request_size;
+    ompi_pml_ucx.request_size     = attr.request_size;
+    ompi_pml_ucx.cuda_initialized = false;
 
+#if HAVE_UCP_ATTR_MEMORY_TYPES && OPAL_CUDA_SUPPORT
```
Comment (yosefe) [inline ompi/mca/pml/ucx/pml_ucx.c:246]: no need for OPAL_CUDA_SUPPORT
use `#if HAVE_UCP_ATTR_MEMORY_TYPES`

Comment (bureddy) [reply ompi/mca/pml/ucx/pml_ucx.c:246]: we want to handle a case where UCX supports cuda but OMPI did not build with cuda.

**Pull request merged**